### ISSUE TITLE
Converting step-8 to RxJava 2

### ISSUE THREAD

**Issue by AshleyByeUK** [vert-x3/vertx-guide-for-java-devs#20]
Using RxJava 2, the following code from [`WikiDatabaseServiceImpl`](https://github.com/vert-x3/vertx-guide-for-java-devs/blob/master/step-8/src/main/java/io/vertx/guides/wiki/database/WikiDatabaseServiceImpl.java#L122-L128) has a bunch of errors due to the change in API:

```
  @Override
  public WikiDatabaseService createPage(String title, String markdown, Handler<AsyncResult<Void>> resultHandler) {
    dbClient.rxUpdateWithParams(sqlQueries.get(SqlQuery.CREATE_PAGE), new JsonArray().add(title).add(markdown))
      .map(res -> (Void) null)
      .subscribe(RxHelper.toSubscriber(resultHandler));
    return this;
  }
```

If I change this to:

```
  @Override
  public WikiDatabaseService createPage(String title, String markdown, Handler<AsyncResult<Void>> resultHandler) {
    dbClient.rxUpdateWithParams(sqlQueries.get(SqlQuery.CREATE_PAGE), new JsonArray().add(title).add(markdown))
      .map(res -> (Void) null)
      .subscribe(SingleHelper.toObserver(resultHandler));
    return this;
  }
```

I get `the mapper function returned a null value`, having forgotten that RxJava 2 does not allow nulls. 

Since the `resultHandler` is expecting a `Handler<AsyncResult<Void>>` type, `Observable.empty()` cannot be used in its place. I've tried converting to `Completable` but also without success.

Based on the given API for the `WikiDatabaseService`, how should the implementation be modified to support RxJava 2?

**Comment by jponge** [comment]
RxJava 2 has `Maybe<T>` to replace the case of `Single` returning `null`.

**Comment by AshleyByeUK** [comment]
Thanks Julien. I'd not considered a `Maybe<T>`. I'm struggling to figure out how to map that to `Void` though, or should I change the method signature in the `WikiDatabaseService` interface to `Handler<AsyncResult<UpdateResult>> resultHandler`?

**Comment by jponge** [comment]
I have yet to investigate upgrading these samples to RxJava 2 so if you are willing to help then you are more than welcome :-)

But I think you can temporarily replace the `map` to `null` by a `toCompletable()` and provide a manual handler if https://github.com/vert-x3/vertx-rx/blob/master/rx-java2/src/main/java/io/vertx/reactivex/CompletableHelper.java does not help.

**Comment by AshleyByeUK** [comment]
More than happy to try and contribute. I've gone back to the original RxJava version and started the process to convert to Rx2 again. Whilst the null is still an issue, I want to try `.map(res -> Observable.<Void>empty())`.

However, I'm actually getting the following error `user lacks privilege or object not found: PAGES in statement [insert into Pages values (NULL, ?, ?)]`, which I believe is an HSQLDB related issue. When I look at the tables, I can't find `Pages`. I suspect this wasn't showing before as I hadn't deleted the database created by the original `CREATE_PAGES_TABLE` when I first ran the original example.

I'm a bit stumped as to how to deal with this or go about diagnosing it. I'll have a scratch of my head and see what I can determine.

**Comment by AshleyByeUK** [comment]
Okay. Cracked it.

First, the issue with lacking privileges. The `Pages` table wasn't being created. This is due to `rxExecute` returning a `Completable` rather than a `Single<Void>`, which also means we can't `flatMap()` the returned value. Instead, we need to use `map()` and subscribe to the `Completable`. We then `map()` the returned `Disposable` to the instance of `WikiDatabaseServiceImpl`. The resulting constructor is as follows:

```
WikiDatabaseServiceImpl(io.vertx.ext.jdbc.JDBCClient dbClient, HashMap<SqlQuery, String> sqlQueries, Handler<AsyncResult<WikiDatabaseService>> readyHandler) {
  this.dbClient = new JDBCClient(dbClient);
  this.sqlQueries = sqlQueries;

  getConnection()
      .map(conn -> conn.rxExecute(sqlQueries.get(SqlQuery.CREATE_PAGES_TABLE)).subscribe())
      .map(disposable -> this)
      .subscribe(SingleHelper.toObserver(readyHandler));
}
```

Secondly, the issue with not being able to push `null` values can be solved by using `subscribe(result, error)` and manually completing the `resultHandler`. As an example, the `createPage()` method is now as follows:

```
@Override
public WikiDatabaseService createPage(String title, String markdown, Handler<AsyncResult<Void>> resultHandler) {
  dbClient
      .rxUpdateWithParams(
          sqlQueries.get(SqlQuery.CREATE_PAGE), new JsonArray().add(title).add(markdown))
      .subscribe(
          res -> resultHandler.handle(Future.succeededFuture()),
          err -> resultHandler.handle(Future.failedFuture(err)));
  return this;
}
```

If I can find the time, I'll clone this example repo, make the changes and submit a pull request. However, I've been working with gradle and `step-9` on its own, so it'll take a bit of time to do.

I hope that's useful?

**Comment by jponge** [comment]
Thanks @AshleyByeUK, looking forward to a pull-request!

**Comment by tsegismont** [comment]
@jponge this was fixed in #22 . Shall we close it?

**Comment by jponge** [comment]
Yep!
